community/codegen/src/main/java/org/neo4j/codegen/CodeBlock.java

```diff
@@ -117,74 +137,82 @@ public Expression self()
 
     public Expression load( String name )
     {
-        return Expression.load( local( name ), name );
+        return Expression.load( local( name ) );
     }
 
+    /*
+     * Foreach is just syntactic sugar for a while loop.
+     *
+     */
     public CodeBlock forEach( Parameter local, Expression iterable )
     {
-        emitter.beginForEach( local, iterable );
-        return new CodeBlock( this );
+        String iteratorName = local.name() + "Iter";
+
+        assign( Iterator.class, iteratorName, Expression.invoke( iterable,
+                MethodReference.methodReference( Iterable.class, Iterator.class, "iterator" ) ) );
+        CodeBlock block = whileLoop( Expression
+                .invoke( load( iteratorName ), methodReference( Iterator.class, boolean.class, "hasNext" ) ) );
+        block.assign( local.type(), local.name(),
+                Expression.cast( local.type(), Expression.invoke( block.load( iteratorName ),
+                        methodReference( Iterator.class, Object.class, "next" ) ) ) );
+
+        return block;
     }
 
     public CodeBlock whileLoop( Expression test )
     {
-        emitter.beginWhile( test );
+        emit( e -> e.beginWhile( test ) );
         return new CodeBlock( this );
     }
 
     public CodeBlock ifStatement( Expression test )
     {
-        emitter.beginIf( test );
+        emit( e -> e.beginIf( test ) );
         return new CodeBlock( this );
     }
 
-    CodeBlock emitCatch( Parameter exception )
+    public CodeBlock ifNotStatement( Expression test )
```

Comment:
I agree we want to use those opcodes. I disagree with this being something we should surface at this level of the API. Instead I think we should specialise the _compare to null_ and _not_ `Expression` subtypes.

The `not` subtype should also specialise so that `not(not(expr))` evaluates to `expr`.